hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/OzoneClientKeyReadWriteOps.java

```diff
@@ -0,0 +1,252 @@
+import org.apache.hadoop.hdds.cli.HddsVersionProvider;
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.ozone.client.OzoneBucket;
+import org.apache.hadoop.ozone.client.OzoneClient;
+import org.apache.hadoop.ozone.client.io.OzoneInputStream;
+import org.apache.hadoop.ozone.client.io.OzoneOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import picocli.CommandLine;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.apache.hadoop.ozone.freon.KeyGeneratorUtil.FILE_DIR_SEPARATOR;
+
+/**
+ * Ozone key generator/reader for performance test.
+ */
+
+@CommandLine.Command(name = "ockrw",
+        aliases = "ozone-client-key-read-write-ops",
+        description = "Read and write keys with the help of the ozone clients.",
+        versionProvider = HddsVersionProvider.class,
+        mixinStandardHelpOptions = true,
+        showDefaultValues = true)
+public class OzoneClientKeyReadWriteOps extends BaseFreonGenerator
+        implements Callable<Void> {
+
+  @CommandLine.Option(names = {"-v", "--volume"},
+          description = "Name of the volume which contains the test data. " +
+                  "Will be created if missing.",
+          defaultValue = "vol1")
+  private String volumeName;
+
+  @CommandLine.Option(names = {"-b", "--bucket"},
+          description = "Name of the bucket which contains the test data.",
+          defaultValue = "bucket1")
+  private String bucketName;
+
+  @CommandLine.Option(names = {"-m", "--read-metadata-only"},
+          description = "If only read key's metadata. " +
+                  "Supported values are Y, F.",
+          defaultValue = "false")
+  private boolean readMetadataOnly;
+  @CommandLine.Option(names = {"-s", "--start-index"},
+          description = "Start index of keys of read/write operation.",
+          defaultValue = "0")
+  private int startIndex;
+
+  @CommandLine.Option(names = {"-r", "--range"},
+          description = "index range of read/write operations.",
+          defaultValue = "0")
+  private int range;
+  @CommandLine.Option(names = {"--size"},
+          description = "Object size (in bytes) " +
+                  "to be generated.",
+          defaultValue = "1")
+  private int objectSizeInBytes;
+
+  @CommandLine.Option(names = {"--keySorted"},
+          description = "Generated sorted key or not. The key name " +
+                  "will be generated via md5 hash if choose " +
+                  "to use unsorted key.",
+          defaultValue = "false")
+  private boolean keySorted;
+
+  @CommandLine.Option(names = {"--percentage-read"},
+          description = "Percentage of read tasks in mix workload.",
+          defaultValue = "100")
+  private int percentageRead;
+
+  @CommandLine.Option(names = {"--clients"},
+          description =
+                  "Number of clients, defaults 1.",
```

Comment:
This is the number of clients instantiated which can equate the number of concurrent TCP sessions to OM or Datanode. The range of keys is partitioned across the clients, and each client generates the read and or write load as per the parameters specified.

Reply:
updated!